Source/mpqapi.cpp

```diff
@@ -614,32 +646,35 @@ bool mpqapi_has_file(const char *pszName)
 
 bool OpenMPQ(const char *pszArchive)
 {
-	_FILEHEADER fhdr;
+	FileHeader fhdr;
 
 	if (!cur_archive.Open(pszArchive)) {
 		return false;
 	}
-	if (cur_archive.sgpBlockTbl == nullptr || cur_archive.sgpHashTbl == nullptr) {
+	if (Cache.name == cur_archive.name) {
+		cur_archive.hashTbl = std::move(Cache.hashTbl);
+		cur_archive.blockTbl = std::move(Cache.blockTbl);
+		Cache = {};
+	} else {
+		Cache = {};
 		if (!cur_archive.exists) {
 			InitDefaultMpqHeader(&cur_archive, &fhdr);
 		} else if (!ReadMPQHeader(&cur_archive, &fhdr)) {
 			goto on_error;
 		}
-		cur_archive.sgpBlockTbl = new _BLOCKENTRY[BlockEntrySize / sizeof(_BLOCKENTRY)];
-		std::memset(cur_archive.sgpBlockTbl, 0, BlockEntrySize);
+		cur_archive.blockTbl.resize(INDEX_ENTRIES);

```

Comment:
This always uses a static size? Can blockTbl be made a std::array instead of a vector?

Reply:
It can be made a `std::array`, but if you want to move the tables quickly to/from the cache (rather than copying them), you'd have to go the extra step and make it a `std::unique_ptr<std::array>>`. This would mess up the syntax when accessing the tables.

Reply:
Rather than a vector for a fixed-size array, use a `unique_ptr<T[]>`. It is specialised for square brackets access.

Reply:
What would the advantage of using `unique_ptr<T[]>` over `vector` be? We lose the ability to do range-based for loops.

Reply:
A vector is not the appropriate type for a fixed-size container. While it offers the nicest syntax among the options in stdlib, that alone not a good reason to use the wrong abstraction. In this case, `unique_ptr<array>` is also a good option.

Reply:
I really don't see how `unique_ptr<T[]>` is superior to `vector` in this case, when expressing and enforcing the container size is the main issue:

- `vector` is resizable when it ideally shouldn't be.
- `unique_ptr<T[]>` has indeterminate size, and can be replaced with an array of a different size.

Reply:
Sounds like `unique_ptr<array>` is the most appropriate type here, despite somewhat awkward access syntax